### ISSUE TITLE
ENH: Another approach np.nan in geometry

### ISSUE THREAD

**Issue by mtomaszewski95** [geopandas/geopandas#1734]
Main idea of this issue is to allow keeping np.nan value while proceeding geopandas functions.
For example:
geopandas.points_from_xy(np.nan, np.nan) = np.nan,
geopandas.points_from_xy(1, np.nan) = np.nan,
geopandas.points_from_xy(np.nan, 1) = np.nan.
It could be done by something like that:

 def _points_from_xy(x, y, z=None):
    # helper method for shapely-based function
    if not len(x) == len(y):
        raise ValueError("x and y arrays must be equal length.")
    if z is not None:
        if not len(z) == len(x):
            raise ValueError("z array must be same length as x and y.")
        geom = [shapely.geometry.Point(i, j, k) if !i.isna() and !j.isna() and !k.isna() else np.nan for i, j, k in zip(x, y, z)]
    else:
        geom = [shapely.geometry.Point(i, j) if !i.isna() and !j.isna() else np.nan  for i, j in zip(x, y)]
    return geom

def points_from_xy(x, y, z=None):

    x = np.asarray(x, dtype="float64")
    y = np.asarray(y, dtype="float64")
    if z is not None:
        z = np.asarray(z, dtype="float64")

    if compat.USE_PYGEOS:
        if x.isna() or y.isna() or z.isna()
              return np.nan
       else
             return pygeos.points(x, y, z)
    else:
        out = _points_from_xy(x, y, z)
        aout = np.empty(len(x), dtype=object)
        aout[:] = out
        return aout

**Comment by martinfleis** [comment]
Thanks for the suggestion. While I understand the motivation, there are reasons why do not support this at the moment. 

`np.nan` is a float. `GeometryArray` supports only geometry objects or `None`. Having a float among geometries is not optimal.

It is the truth that there's no easy way to check if coordinates of geometry are `np.nan` so once it happens, it is hard to filter these potentially erroneous geometries out (although they are not valid and `is_valid` returns `False`).

I'd say that optimal would be a method which would allow to filter them. Something like `has_nan` (better name needed) which would return a boolean array. Then you could replace geometries with `np.nan` either with `None` or drop rows.

Would that work for your use case?

**Comment by mtomaszewski95** [comment]
The way You suggest is probably equivalent to left handling user with np.nan. Now, if proceed points_from_xy in pd.DataFrame, that contains row with both coorindates np.nan then gets point(nan nan). So user could then replace this value with np.nan with np.where function.
If geopandas will not support np.nan, then better option is to automatically recognize potential output (like point(nan nan)) and replace this with Non, so when user proceed functions from library, he gets more expected values.

**Comment by olsgaard** [comment]
> `GeometryArray` supports only geometry objects or `None`.

It's not really `None`, but `EMPTY POINT`, right? And that isn't understood as NaN/None by the rest of the pandas ecosystem, leading to problems like:

```python
s1 = gpd.GeoSeries(gpd.points_from_xy([1,2, None, None], [1,2,None, None]))
s2 = gpd.GeoSeries(index=[2,3], data=[Point(4,5),Point(5,6)])

print(s1)
print(s1.isna())
print(s1.fillna(s2))
```

Expected result
```
0    POINT (1.00000 1.00000)
1    POINT (2.00000 2.00000)
2                POINT EMPTY
3                POINT EMPTY
dtype: geometry
0    False
1    False
2     True
3     True
dtype: bool
0    POINT (1.00000 1.00000)
1    POINT (2.00000 2.00000)
2    POINT (4.00000 5.00000)
3    POINT (5.00000 6.00000)
dtype: geometry
```

Actual result
```
0    POINT (1.00000 1.00000)
1    POINT (2.00000 2.00000)
2                POINT EMPTY
3                POINT EMPTY
dtype: geometry
0    False
1    False
2    False
3    False
dtype: bool
0    POINT (1.00000 1.00000)
1    POINT (2.00000 2.00000)
2                POINT EMPTY
3                POINT EMPTY
dtype: geometry
```

**Comment by martinfleis** [comment]
> It's not really None, but EMPTY POINT, right? 

No, it is really `None`. But passing `None` as **coordinates** to `points_from_xy` create an empty point.
